File: backend/cmd/gamelab/monopoly.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	mono "github.com/agent-arena/arena/internal/engine/monopoly"
)

// Monopoly for the lab agent.
//
// # Why this file had to exist
//
// The lab advertised `supportedGames: ["goofspiel", "mafia", "monopoly"]` and implemented one of
// them. A Monopoly view hit the default branch of handlePlay, which answered `{}` and let the
// platform apply its legal fallback — so every lab Monopoly match was the SERVER playing itself
// while the logs read as though four agents were competing. Nothing errored, which is what made
// it survive: a forfeit and a decision are both a move on the wire.
//
// That is the same shape as the instrumentation bug where the driver was measured instead of the
// match, and it matters most right before a stress run. A load test against fallbacks measures
// the fallback path at full speed and reports it as agent throughput.
//
// # Why it reads the real board
//
// The prices, groups and rents come from mono.Board() rather than a table copied into the lab.
// A private copy is wrong the first time the engine rebalances anything, and it would be wrong
// silently — the agent would keep bidding confident numbers against a board that had moved.
// monopolyView MUST mirror internal/monopoly.MonopolyPushView field for field.
//
// It did not, and the cost was total: this struct read `your_seat` and `legal` while the
// platform sends `seat` and `legal_actions`. Every field silently decoded to its zero value,
// so `len(v.Legal) == 0` was true on every turn, the handler returned `{}` before it ever
// logged, and the platform's safeFallback played BOTH seats of every lab Monopoly match. A
// full match ran to GAME END with zero decision lines in the agent log.
//
// That is precisely the failure this file's header describes and claims to have fixed — it
// was fixed against the OLD wire shape, and the push protocol renamed the fields. The seat
// number is the worst of them: v.YourSeat indexes st.Players, so seat 1 would have evaluated
// the board as if it were seat 0 even had the rest decoded.
//
// monopolyWireTest pins this struct against the platform's own type. Do not rename a field
// here without renaming it there.
type monopolyView struct {
	MatchID string `json:"match_id"`
	// Seat, not your_seat. Indexes State.Players — a wrong value reads someone else's board.
	YourSeat int    `json:"seat"`
	Phase    string `json:"phase"`
	// legal_actions, not legal. An empty list means "no decision to make", so a name that
	// never matches is indistinguishable from a seat that has nothing to do.
	Legal []string    `json:"legal_actions"`
	State *mono.State `json:"state"`
	// Round is the engine's TurnCount, published by the platform so both sides agree on the
	// number the turn proof was minted for. Without it a bound call cannot verify.
	Round int `json:"round"`
	// TurnProof binds a gateway model call to THIS decision.
	TurnProof string `json:"turn_proof"`
}

func (a *labAgent) playMonopoly(w http.ResponseWriter, r *http.Request, raw []byte) {
	var v monopolyView
	if err := json.Unmarshal(raw, &v); err != nil || v.State == nil {
		http.Error(w, "bad view", http.StatusBadRequest)
		return
	}
	if len(v.Legal) == 0 {
		// Nothing is legal for this seat right now. An empty object is correct here and is NOT
		// the silent fallback this file replaced: there is genuinely no decision to make.
		writeJSON(w, map[string]any{})
		return
	}

	// GO DARK, same contract as Goofspiel: hold the connection open rather than erroring, so the
	// shot clock actually expires and the absence forfeit arms. An error would be answered
	// instantly and would exercise the wrong path.
	if GoDarkAfterRound > 0 && v.State.RollSeq >= GoDarkAfterRound &&
		(GoDarkSeat < 0 || GoDarkSeat == v.YourSeat) {
		a.log.Printf("monopoly: seat %d GONE DARK at roll %d", v.YourSeat, v.State.RollSeq)
		<-r.Context().Done()
		return
	}

	think := a.Persona.thinkTime(v.MatchID, v.State.RollSeq, v.YourSeat)
	time.Sleep(think)

	// BOUND PATH: ask the model through the gateway so the decision is completion-bound,
	// making this a real LLM agent rather than a scripted persona. Same contract as Mafia:
	// ANY failure falls back to the rule policy and SAYS SO, because a run that silently
	// played a scripted move while reporting a bound one measures nothing and claims success.
	var act map[string]any
	var why string
	if BindGatewayBase != "" {
		me := v.State.Players[v.YourSeat]
		prompt := fmt.Sprintf(
			"You are seat %d in Monopoly, phase %s, turn %d. You hold $%d.",
			v.YourSeat, v.Phase, v.Round, me.Cash)
		b, err := a.decideGameThroughGateway("monopoly", v.MatchID, v.Round, v.YourSeat,
			v.TurnProof, v.Legal, prompt)
		if err != nil {
			a.log.Printf("monopoly: seat %d BIND FAILED (%v) — falling back to the rule policy, "+
				"so THIS turn is not model-backed", v.YourSeat, err)
		} else {
			act = map[string]any{"kind": b.Kind}
			if b.Property != 0 {
				act["property"] = b.Property
			}
			if b.Amount != 0 {
				act["amount"] = b.Amount
			}
			why = "bound to the model's own " + b.Canon
		}
	}
	if act == nil {
		act, why = monopolyAction(a.Persona, v)
	}
	a.log.Printf("monopoly: seat %d  phase %-8s legal %v  → %s (%s)",
		v.YourSeat, v.Phase, v.Legal, act["kind"], why)

	act["rationale"] = why
	act["usage"] = a.usage(len(raw), think)
	writeJSON(w, monopolyWireMove(act))
}

// monopolyWireMove renames the engine-shaped action to what MonopolyPushMove reads.
//
// The lab builds actions with the engine's own vocabulary — `kind`, matching mono.Action —
// but the push protocol's move field is `action`. Sending `kind` meant Action decoded empty,
// the platform found it not in the legal list, and substituted a safe fallback: the second
// half of the same silent-fallback bug as the view above, and equally invisible because an
// illegal move and a considered one are both just a move on the wire.
//
// Translating here rather than renaming the key throughout keeps monopolyAction and its tests
// speaking the engine's language, which is what they are actually about.
func monopolyWireMove(act map[string]any) map[string]any {
	out := make(map[string]any, len(act))
	for k, v := range act {
		if k == "kind" {
			out["action"] = v
			continue
		}
		out[k] = v
	}
	return out
}

// monopolyAction picks a move. Returns the action object and the rationale the decision
// inspector shows.
//
// Every branch chooses from v.Legal and never invents a verb: the engine is the authority on
// what is playable, and a lab agent that guesses would be testing our guess rather than the
// contract an SDK agent has to satisfy.
func monopolyAction(p persona, v monopolyView) (map[string]any, string) {
	st := v.State
	me := st.Players[v.YourSeat]
	board := mono.Board()
	legal := legalSet(v.Legal)

	switch {
	// ── debt: raise cash or go under ─────────────────────────────────────────
	// Checked FIRST because a seat in debt may also have `build` legal, and building while
	// insolvent is how an agent bankrupts itself with money in its pocket.
	case legal[mono.ActSellHouse] || legal[mono.ActMortgage] || legal[mono.ActBankrupt]:
		owed := 0
		if st.Debt != nil {
			owed = st.Debt.Amount
		}
		if me.Cash >= owed && legal[mono.ActEndTurn] {
			return map[string]any{"kind": mono.ActEndTurn}, "debt already covered"
		}
		// Mortgage the cheapest undeveloped holding first: it raises cash while giving up the
		// least rent, and mortgaging a developed street would waste the houses on it.
		if legal[mono.ActMortgage] {
			if idx, ok := cheapestMortgageable(st, board, v.YourSeat); ok {
				return map[string]any{"kind": mono.ActMortgage, "property": idx},
					"mortgaging the cheapest undeveloped holding to cover the debt"
			}
		}
		if legal[mono.ActSellHouse] {
			if idx, ok := mostDevelopedOwned(st, board, v.YourSeat); ok {
				return map[string]any{"kind": mono.ActSellHouse, "property": idx},
					"selling a house — nothing left to mortgage"
			}
		}
		return map[string]any{"kind": mono.ActBankrupt}, "cannot raise the debt; conceding"

	// ── jail ─────────────────────────────────────────────────────────────────
	case legal[mono.ActUseJailCard]:
		return map[string]any{"kind": mono.ActUseJailCard}, "spending a held jail card — it costs nothing"
	case legal[mono.ActPayJail] || legal[mono.ActRollJail]:
		// Early game, get out and keep buying. Late game, jail is shelter: rent is the main
		// risk once the board is developed, and a jailed player pays none.
		developed := developedCount(st)
		if legal[mono.ActPayJail] && me.Cash > 200 && developed < 8 {
			return map[string]any{"kind": mono.ActPayJail}, "paying out early — board is still open to buy"
		}
		if legal[mono.ActRollJail] {
			return map[string]any{"kind": mono.ActRollJail}, "rolling for doubles; jail is cheap shelter on a developed board"
		}
		return map[string]any{"kind": mono.ActPayJail}, "paying the fine"

	// ── acquire ──────────────────────────────────────────────────────────────
	case legal[mono.ActBuy] || legal[mono.ActDecline]:
		sp := board[me.Position]
		if legal[mono.ActBuy] && wantsToBuy(p, st, board, v.YourSeat, me.Position) {
			return map[string]any{"kind": mono.ActBuy}, "buying " + sp.Name + " at list"
		}
		return map[string]any{"kind": mono.ActDecline}, "declining " + sp.Name + "; cash reserve matters more"

	// ── auction ──────────────────────────────────────────────────────────────
	case legal[mono.ActBid] || legal[mono.ActPass]:
		if st.Auction == nil {
			return map[string]any{"kind": mono.ActPass}, "no auction state; passing"
		}
		sp := board[st.Auction.Property]
		ceiling := bidCeiling(p, st, board, v.YourSeat, st.Auction.Property)
		next := st.Auction.HighBid + 10
		if legal[mono.ActBid] && next <= ceiling && next <= me.Cash {
			// Always a positive amount. An amount-less bid is exactly the mistake that used to
			// come back as "not legal in the current phase", and the lab must not reproduce it.
			return map[string]any{"kind": mono.ActBid, "amount": next},
				"raising on " + sp.Name + " — still under my valuation"
		}
		return map[string]any{"kind": mono.ActPass}, "passing on " + sp.Name + " — bidding past my valuation"

	// ── trade: open floor ────────────────────────────────────────────────────
	case legal[mono.ActSkipTrade]:
		// A proposal needs a full Trade payload; offering `propose_trade` with none is rejected,
		// which is precisely what the first version of this file did via its "first legal verb"
		// fallback. Propose only when a concrete, well-formed offer exists.
		if legal[mono.ActProposeTrade] {
			if tr, why, ok := buildTradeOffer(st, board, v.YourSeat); ok {
				return map[string]any{"kind": mono.ActProposeTrade, "trade": tr}, why
			}
		}
		return map[string]any{"kind": mono.ActSkipTrade}, "no offer worth making this window"

	// ── trade: responding ────────────────────────────────────────────────────
	case legal[mono.ActAcceptTrade] || legal[mono.ActRejectTrade]:
		if st.PendingTrade != nil && tradeIsGood(st, board, v.YourSeat, st.PendingTrade) && legal[mono.ActAcceptTrade] {
			return map[string]any{"kind": mono.ActAcceptTrade}, "accepting — the offer nets me a group I want"
		}
		return map[string]any{"kind": mono.ActRejectTrade}, "rejecting — the offer favours the proposer"

	// ── manage / roll ────────────────────────────────────────────────────────
	case legal[mono.ActBuild]:
		if idx, ok := bestBuild(st, board, v.YourSeat, me.Cash); ok {
			return map[string]any{"kind": mono.ActBuild, "property": idx},
				"building on " + board[idx].Name + " — completed group, cash to spare"
		}
		if legal[mono.ActEndTurn] {
			return map[string]any{"kind": mono.ActEndTurn}, "holding cash rather than over-building"
		}
	}

	// Prefer rolling over ending a turn: a roll is the only action that advances the game, and a
	// lab that ends its turn whenever both are legal produces matches that never resolve.
	if legal[mono.ActRoll] {
		return map[string]any{"kind": mono.ActRoll}, "rolling"
	}
	if legal[mono.ActEndTurn] {
		return map[string]any{"kind": mono.ActEndTurn}, "nothing worth doing this turn"
	}
	// Last resort for a phase this file has not been taught.
	//
	// Deliberately NOT "the first legal verb". That was the original fallback and it produced an
	// ILLEGAL action: in the trade window the first legal verb is propose_trade, which is only
	// valid with a Trade payload attached. A fallback that reaches for a verb needing arguments
	// it cannot supply turns an unknown phase into a rejected move.
	//
	// So: prefer the verbs that are complete on their own — declining, passing, ending — and only
	// then take whatever is left.
	for _, safe := range []string{mono.ActSkipTrade, mono.ActRejectTrade, mono.ActPass,
		mono.ActDecline, mono.ActEndTurn, mono.ActRoll} {
		if legal[safe] {
			return map[string]any{"kind": safe}, "unhandled phase " + v.Phase + "; taking the no-argument action " + safe
		}
	}
	return map[string]any{"kind": v.Legal[0]}, "unhandled phase " + v.Phase + "; no argument-free action available"
}

func legalSet(l []string) map[string]bool {
	m := make(map[string]bool, len(l))
	for _, k := range l {
		m[k] = true
	}
	return m
}

// wantsToBuy: take anything that completes or extends a group, and otherwise buy while a cash
// buffer survives the purchase. The buffer is what stops an agent owning the board and losing to
// the first rent it lands on.
func wantsToBuy(p persona, st *mono.State, board []mono.Space, seat, pos int) bool {
	sp := board[pos]
	if sp.Price == 0 || st.Players[seat].Cash < sp.Price {
		return false
	}
	reserve := 150
	if p.Style == "aggressive" {
		reserve = 60
	}
	if groupInterest(st, board, seat, sp.Group) > 0 {
		reserve /= 2 // already invested here; completing a group is worth more than the buffer
	}
	return st.Players[seat].Cash-sp.Price >= reserve
}

// bidCeiling values a property at list, plus a premium when it completes a group we already hold
// and a discount when a rival is closer to completing it than we are.
func bidCeiling(p persona, st *mono.State, board []mono.Space, seat, pos int) int {
	sp := board[pos]
	if sp.Price == 0 {
		return 0
	}
	ceiling := sp.Price
	if mine := groupInterest(st, board, seat, sp.Group); mine > 0 {
		ceiling = sp.Price * (100 + 40*mine) / 100
	}
	if p.Style == "aggressive" {
		ceiling = ceiling * 120 / 100
	}
	// Never bid past the cash on hand: the engine rejects it as insufficient funds, and an agent
	// that keeps proposing rejected bids stalls its own auction.
	if cash := st.Players[seat].Cash; ceiling > cash {
		ceiling = cash
	}
	return ceiling
}

// groupInterest counts how many squares of a colour group this seat already owns.
func groupInterest(st *mono.State, board []mono.Space, seat int, group string) int {
	if group == "" {
		return 0
	}
	n := 0
	for i, sp := range board {
		if sp.Group == group && i < len(st.Holdings) && st.Holdings[i].Owner == seat {
			n++
		}
	}
	return n
}

// bestBuild picks the square to develop: the completed, unmortgaged group with the fewest houses,
// so development spreads evenly instead of stacking one street (which the even-build rule forbids
// anyway, and which would have the engine reject the move).
func bestBuild(st *mono.State, board []mono.Space, seat, cash int) (int, bool) {
	best, bestHouses := -1, 99
	for i, sp := range board {
		if sp.Kind != mono.KindStreet || i >= len(st.Holdings) {
			continue
		}
		h := st.Holdings[i]
		if h.Owner != seat || h.Mortgaged || h.Houses >= 5 {
			continue
		}
		if groupInterest(st, board, seat, sp.Group) < groupSize(board, sp.Group) {
			continue // partial group: cannot build, and asking would be rejected
		}
		if cash-sp.HouseCost < 100 {
			continue // keep a rent buffer; a house is worthless if the next square bankrupts us
		}
		if h.Houses < bestHouses {
			best, bestHouses = i, h.Houses
		}
	}
	return best, best >= 0
}

func groupSize(board []mono.Space, group string) int {
	n := 0
	for _, sp := range board {
		if sp.Group == group {
			n++
		}
	}
	return n
}

// cheapestMortgageable: raise cash with the smallest loss of rent.
func cheapestMortgageable(st *mono.State, board []mono.Space, seat int) (int, bool) {
	best, bestPrice := -1, 1<<30
	for i, sp := range board {
		if i >= len(st.Holdings) || !sp.Ownable() {
			continue
		}
		h := st.Holdings[i]
		if h.Owner != seat || h.Mortgaged || h.Houses > 0 {
			continue // houses must be sold before a street can be mortgaged
		}
		if sp.Price < bestPrice {
			best, bestPrice = i, sp.Price
		}
	}
	return best, best >= 0
}

func mostDevelopedOwned(st *mono.State, board []mono.Space, seat int) (int, bool) {
	best, bestHouses := -1, 0
	for i := range board {
		if i >= len(st.Holdings) {
			continue
		}
		if h := st.Holdings[i]; h.Owner == seat && h.Houses > bestHouses {
			best, bestHouses = i, h.Houses
		}
	}
	return best, best >= 0
}

// developedCount is a rough measure of how far along the board is, used to decide whether jail is
// a cost or a shelter.
func developedCount(st *mono.State) int {
	n := 0
	for _, h := range st.Holdings {
		n += h.Houses
	}
	return n
}

// buildTradeOffer constructs a complete, well-formed offer, or reports that none is worth making.
//
// Returns ok=false rather than a half-filled Trade: the engine validates the payload, and an
// offer naming properties the proposer does not own is a rejected move, not a bad negotiation.
func buildTradeOffer(st *mono.State, board []mono.Space, seat int) (mono.Trade, string, bool) {
	me := st.Players[seat]
	if me.Cash < 200 {
		return mono.Trade{}, "", false // no cash to sweeten anything
	}
	// Look for a square that would COMPLETE a group for us and is owned by someone else.
	for i, sp := range board {
		if sp.Kind != mono.KindStreet || i >= len(st.Holdings) {
			continue
		}
		h := st.Holdings[i]
		if h.Owner == seat || h.Owner == mono.Bank || h.Houses > 0 {
			continue
		}
		if owner := h.Owner; owner >= 0 && owner < len(st.Players) && !st.Players[owner].Bankrupt {
			held := groupInterest(st, board, seat, sp.Group)
			if held == 0 || held+1 < groupSize(board, sp.Group) {
				continue // only chase the square that finishes the set
			}
			offer := sp.Price * 3 / 2
			if offer > me.Cash-100 {
				continue // would leave nothing to develop the group we just completed
			}
			return mono.Trade{
				Proposer: seat, Target: h.Owner,
				GiveCash: offer, WantProps: []int{i},
			}, "offering " + itoa(offer) + " for " + sp.Name + " — it completes my group", true
		}
	}
	return mono.Trade{}, "", false
}

// tradeIsGood accepts only offers that leave us better off: cash above the list price of what we
// give up, or a square that completes one of our own groups.
func tradeIsGood(st *mono.State, board []mono.Space, seat int, tr *mono.Trade) bool {
	if tr.Target != seat {
		return false
	}
	giveValue := 0
	for _, i := range tr.WantProps {
		if i >= 0 && i < len(board) {
			giveValue += board[i].Price
			if groupInterest(st, board, seat, board[i].Group) >= groupSize(board, board[i].Group) {
				return false // never break up a completed group
			}
		}
	}
	gainValue := tr.GiveCash
	for _, i := range tr.GiveProps {
		if i >= 0 && i < len(board) {
			gainValue += board[i].Price
		}
	}
	return gainValue > giveValue*5/4 // needs a clear premium, not a coin flip
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var b []byte
	for n > 0 {
		b = append([]byte{byte('0' + n%10)}, b...)
		n /= 10
	}
	if neg {
		return "-" + string(b)
	}
	return string(b)
}

File: harness/bench/pyyolbench/games/monopoly.py
"""Monopoly: near-perfect information, and by far the most expensive arena to benchmark.

# Read this before running Monopoly on a frontier model

A Monopoly match can run to the turn cap — hundreds of turns across four seats, each turn
several phases, each phase a model call. Where a Goofspiel match is 26 decisions, a Monopoly
match can be thousands. At frontier prices that is the difference between cents and tens of
dollars for ONE match. The run plan therefore treats Monopoly as the arena you enter last,
with the cheapest viable models and a hard turn cap, and the $10 pilot does not enter it at
all. This is a cost property of the game, not a limitation of the harness.

# Why the prompt says "read legal_actions" more than once

Monopoly is a phase machine and the legal set already encodes affordability, even-build rules
and whose turn it is. Any action in that list is guaranteed to be accepted; anything outside
it is guaranteed to be refused. So the whole legality question reduces to "choose from the
list", and a model that instead reasons from remembered Monopoly rules will propose builds it
cannot afford on sets it does not own. Stating this plainly is worth more than any strategy
paragraph below it.
"""

from __future__ import annotations

from typing import Any

GAME = "monopoly"

_SYSTEM = """\
You are an expert Monopoly player competing in the Pyyol arena against other AI agents. \
Play to win: be the last solvent player, or hold the highest net worth if the turn cap \
arrives first.

## The shape of a turn

Monopoly here is a phase machine and your view names the phase and the exact actions legal \
right now:

- `roll` — your turn begins; roll the dice.
- `jail` — you are in jail: `pay_jail` ($50 then roll), `use_jail_card`, or `roll_jail` for doubles.
- `acquire` — you landed on an unowned property: `buy` at list price, or `decline` (which \
sends it to auction).
- `auction` — someone declined a property: `bid` with an `amount`, or `pass`.
- `resolve_debt` — you owe more than your cash: `mortgage`, `sell_house`, trade your way out, \
or `bankrupt`.
- `manage` — after moving: `build`, `mortgage`, `unmortgage`, `propose_trade`, then `end_turn`.
- `trade_response` — someone offered you a trade: `accept_trade`, `reject_trade`, `counter_trade`.
- `trade` — an open trade window: `propose_trade` or `skip_trade`.

**Always choose your action from the `legal_actions` list in the view.** That list already \
accounts for your cash, the even-build rule, and whose turn it is. Anything in it will be \
accepted; anything outside it will be refused and cost you the decision. `end_turn` during \
`manage` is always safe if nothing else is worth doing.

## How to play well

1. **Monopolies are the entire game.** Unimproved property barely pays. A completed colour \
group with houses is where nearly all rent comes from. Every decision should be measured \
against "does this bring me closer to a set, or stop an opponent completing one".

2. **Build to three houses, quickly.** The rent jump from two houses to three is the largest \
in the game. Getting one set to three houses beats spreading cash thinly over several.

3. **The orange and red groups are the best real estate**, because players leaving jail land \
there most often. Light blues are the best cheap investment. Utilities are close to worthless; \
railroads are respectable early and stop mattering later.

4. **Houses are a finite resource — denying them is a real strategy.** If houses are scarce, \
buying them keeps opponents from improving at all. A contested house goes to auction here.

5. **Keep a cash buffer.** Being unable to pay rent forces mortgaging at a loss, or bankruptcy. \
Roughly the largest rent on the board is the buffer to hold once opponents have sets.

6. **Auctions are where value is won.** A property nobody wants goes cheap, and cheap property \
that blocks a set is worth well over list. You may raise cash mid-auction. Do not bid up a \
property you do not want just to hurt someone — you may be left holding it.

7. **Trade, but never hand over the last piece of someone's set** unless you get a set of your \
own in return. A trade that gives both sides a monopoly favours whoever has more cash to build. \
Offering to the whole table with target -1 finds the best price when you simply want a property \
sold.

8. **Mortgage strategically, not desperately.** Mortgaging property outside your sets to build \
inside them is good. Mortgaging your own set to survive is the beginning of losing.

## Your output

Call the `take_action` tool with the action verb from `legal_actions`, plus `property` (a board \
square index) and `amount` (a cash figure) when the action needs them, and 0 when it does not. \
Do not answer in prose."""


class MonopolyPolicy:
    game = GAME

    def system_prompt(self) -> str:
        return _SYSTEM

    def render_state(self, view: Any, memory: str) -> str:
        raw = getattr(view, "raw", {}) or {}
        seat = int(getattr(view, "seat", raw.get("seat", 0)) or 0)
        phase = str(getattr(view, "phase", raw.get("phase", "")) or "")
        legal = [str(a) for a in (getattr(view, "legal_actions", None) or raw.get("legal_actions") or [])]
        state = getattr(view, "state", None) or raw.get("state") or {}

        lines = [f"PHASE `{phase}`. You are seat {seat}."]
        lines.append(self._players_table(state, seat))
        holdings = self._holdings(state, seat)
        if holdings:
            lines.append("")
            lines.append(holdings)
        pending = self._pending(state)
        if pending:
            lines.append("")
            lines.append(pending)

        if memory:
            lines.append("")
            lines.append("Your notes from earlier in this match:")
            lines.append(memory)

        lines.append("")
        lines.append(f"Legal actions RIGHT NOW: {legal}")
        lines.append("Call `take_action` with one of exactly these verbs.")
        return "\n".join(lines)

    def _players_table(self, state: dict[str, Any], seat: int) -> str:
        players = state.get("players")
        if not isinstance(players, list) or not players:
            return "Player state unavailable."
        rows = ["Players (cash | position | jail | bankrupt):"]
        for i, p in enumerate(players):
            if not isinstance(p, dict):
                continue
            tag = " ← you" if i == seat else ""
            rows.append(
                f"  seat {i}: ${p.get('cash','?')} | sq {p.get('position','?')}"
                f" | {'JAILED' if p.get('in_jail') or p.get('jail') else 'free'}"
                f" | {'BANKRUPT' if p.get('bankrupt') else 'solvent'}{tag}"
            )
        return "\n".join(rows)

    def _holdings(self, state: dict[str, Any], seat: int) -> str:
        """Only OUR squares, plus a count of everyone else's.

        The full holdings map is 40 squares of nested detail on every single decision. Sent
        whole it would dominate the prompt, push past the cached prefix and re-bill the
        rules each turn. Ours in full and theirs as a count keeps the decision-relevant
        part and drops the bulk.
        """
        holdings = state.get("holdings")
        if not isinstance(holdings, (list, dict)):
            return ""
        items = holdings.items() if isinstance(holdings, dict) else enumerate(holdings)
        mine: list[str] = []
        others: dict[int, int] = {}
        for sq, h in items:
            if not isinstance(h, dict):
                continue
            owner = h.get("owner")
            if owner is None or (isinstance(owner, int) and owner < 0):
                continue
            try:
                owner_i = int(owner)
            except (TypeError, ValueError):
                continue
            if owner_i == seat:
                bits = [f"sq {sq}"]
                if h.get("houses"):
                    bits.append(f"{h['houses']} houses")
                if h.get("mortgaged"):
                    bits.append("MORTGAGED")
                mine.append(" ".join(bits))
            else:
                others[owner_i] = others.get(owner_i, 0) + 1
        out = []
        if mine:
            out.append("Your properties: " + "; ".join(mine))
        else:
            out.append("You own no properties yet.")
        if others:
            out.append("Opponent holdings: " + ", ".join(f"seat {k}: {v} squares" for k, v in sorted(others.items())))
        return "\n".join(out)

    def _pending(self, state: dict[str, Any]) -> str:
        out = []
        auction = state.get("auction") or state.get("pending_auction")
        if isinstance(auction, dict) and auction:
            out.append(
                f"Open auction: square {auction.get('property','?')}, "
                f"high bid ${auction.get('high_bid', auction.get('bid', 0))} "
                f"by seat {auction.get('high_bidder', auction.get('bidder','?'))}"
            )
        trade = state.get("trade") or state.get("pending_trade")
        if isinstance(trade, dict) and trade:
            out.append(f"Pending trade: {trade}")
        return "\n".join(out)

    # --- legality -------------------------------------------------------------------

    def _legal(self, view: Any) -> list[str]:
        raw = getattr(view, "raw", {}) or {}
        return [str(a) for a in (getattr(view, "legal_actions", None) or raw.get("legal_actions") or [])]

    def fallback(self, view: Any) -> dict[str, Any]:
        """The safest legal action that keeps the match moving.

        Preference order is chosen to be inert rather than good: end the turn, decline to
        spend, pass an auction. The one exception is `roll`, which is not a strategic
        choice at all — a phase that only permits rolling has no decision in it, and
        refusing to roll would wedge the match.
        """
        legal = self._legal(view)
        for a in ("roll", "end_turn", "skip_trade", "pass", "decline", "reject_trade"):
            if a in legal:
                return {"action": a, "property": 0, "amount": 0}
        return {"action": legal[0] if legal else "end_turn", "property": 0, "amount": 0}

    def coerce(self, args: dict[str, Any], view: Any) -> tuple[dict[str, Any], str]:
        legal = self._legal(view)
        kind = str(args.get("kind", "") or "").strip().lower()
        if kind not in legal:
            return self.fallback(view), "repair"

        def as_int(key: str) -> int:
            try:
                return int(args.get(key) or 0)
            except (TypeError, ValueError):
                return 0

        move: dict[str, Any] = {
            "action": kind,
            "property": as_int("property"),
            "amount": as_int("amount"),
        }
        # The trade payload is passed through untouched when present. It is deliberately
        # NOT part of the canonical bound form — a nested structure re-rendered cosmetically
        # differently would reject an honest turn — so nothing here can cost the turn its
        # binding, and a malformed trade is refused by the engine rather than by us.
        trade = args.get("trade")
        if kind in ("propose_trade", "counter_trade") and isinstance(trade, dict):
            move["trade"] = trade
        return move, "model"
